cluster/src/dunit/scala/org/apache/spark/sql/CachedBatchScanDUnitTest.scala

```diff
@@ -0,0 +1,124 @@
+          "group by UniqueCarrier order by arrivalDelay")
+
+    df_allCachedBatchesScan.count
+
+    var executionIds =
+      snc.sharedState.listener.executionIdToData.keySet.diff(previousExecutionIds)
+
+    var executionId = executionIds.head
+
+    val (scanned1, skipped1) =
+      findCachedBatchStats(df_allCachedBatchesScan, snc.snappySession, executionId)
+    assert(skipped1 == 0, "All Cached batches should have been scanned")
+    assert(scanned1 > 0, "All Cached batches should have been scanned")
+
+    // ***Check for the case when all the cached batches are skipped****
+    previousExecutionIds = snc.sharedState.listener.executionIdToData.keySet
+
+    val df_noCachedBatchesScan = snc.sql(
+      "select AVG(ArrDelay) arrivalDelay, UniqueCarrier carrier " +
+          "from AIRLINE where ArrDelay > 101  " +
+          "group by UniqueCarrier order by arrivalDelay")
+
+    df_noCachedBatchesScan.count
+
+    executionIds =
+        snc.sharedState.listener.executionIdToData.keySet.diff(previousExecutionIds)
+
+    executionId = executionIds.head
+
+    val (scanned2, skipped2) =
+      findCachedBatchStats(df_allCachedBatchesScan, snc.snappySession, executionId)
+    assert(scanned2 == skipped2, "No Cached batches should have been scanned")
+    assert(skipped2 > 0, "No Cached batches should have been scanned")
+
+    // ***Check for the case when some of the cached batches are scanned ****
+    previousExecutionIds = snc.sharedState.listener.executionIdToData.keySet
+
+    val df_someCachedBatchesScan = snc.sql(
+      "select AVG(ArrDelay) arrivalDelay, UniqueCarrier carrier " +
+          "from AIRLINE where ArrDelay < 20  " +
+          "group by UniqueCarrier order by arrivalDelay")
+
+    df_someCachedBatchesScan.count
+
+    executionIds =
+        snc.sharedState.listener.executionIdToData.keySet.diff(previousExecutionIds)
+
+    executionId = executionIds.head
+
+    val (scanned3, skipped3) =
+      findCachedBatchStats(df_allCachedBatchesScan, snc.snappySession, executionId)
+
+    assert(skipped3 > 0, "Some Cached batches should have been scanned")
+    assert(scanned3 != skipped3, "Some Cached batches should have been scanned - comparison")
+  }
+
+  private def findCachedBatchStats(df: DataFrame,
+      sc: SnappySession, executionId: Long): (Long, Long) = {
+
+    val metricValues = sc.sharedState.listener.getExecutionMetrics(executionId)
+    var a = (sc.sharedState.listener.getRunningExecutions ++
+        sc.sharedState.listener.getCompletedExecutions ).filter(x => {
+      x.executionId == executionId
+    })
+    val seenid = a.head.accumulatorMetrics.filter(x =>
+    {x._2.name == "cached batches seen"}).head._1
+    val skippedid = a.head.accumulatorMetrics.filter(x =>
+    {x._2.name == "cached batches skipped by the predicate"}).head._1
+
+    (metricValues.filter(_._1 == seenid).head._2.toInt,
+        metricValues.filter(_._1 == skippedid).head._2.toInt)
+  }
+
+
```

Comment:
Any reason to have a dunit and not a scalatest? DUnits increase the precheckin runtime by a whole lot and should be avoided unless required. This case looks perfect white-box checks for a scalatest.

For example take a look at Spark test suite. When I had last checked, it had 5X or so more tests than snappy+gemxd+gem combined and still runs in under 2 hours (as compared to >12 hours for all combined for us). IMO we need to be careful about more test coverage and avoid dunits unless required. Likewise we have many tests, especially in AQP, that repeatedly load the same set of airline data taking a whole lot of time but hardly providing proportionate coverage. Had mentioned this to AQP team and they are tracking as an issue.